### PR TITLE
Dispatch to old Jordan-Wigner function from new one

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
There are now two functions called `jordan_wigner` in PennyLane. The existing `jordan_wigner` function is available as `qml.qchem.jordan_wigner` but not imported at top level. The new one lives in `qml.fermi.jordan_wigner`. We want to only have one, imported at top level, but we don't want to break anything for people using the current one in the next PL release, as there is no time for a deprecation cycle.

Additionally, we don't have time to update the many functions that use `jordan_wigner` to follow the new standard for before the release, so we can't add a deprecation warning yet; this would cause other unrelated functions to raise deprecation warnings through no fault of the user.

**Description of the Change:**
The new `jordan_wigner` function uses single dispatch to call one of 3 functions, depending on whether the `fermi_operator` passed to it is a `FermiWord`, `FermiSentence` or `list` (the legacy version). The new `jordan_wigner` function thus encompasses all behaviour of the old one (renamed `_jordan_wigner_legacy`).

The imports have been updated so that the new `jordan_wigner` function can be imported as `qml.jordan_wigner`, `qml.fermi.jordan_wigner`, and `qml.qchem.jordan_wigner`.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -28,8 +28,10 @@
   [(#4191)](https://github.com/PennyLaneAI/pennylane/pull/4191)
 
 * Added a conversion function `jordan_wigner` that converts a fermionic operator (`FermiWord`) to a qubit 
-  operator (`PauliSentence`) using the Jordan-Wigner mapping.
+  operator (`PauliSentence`) using the Jordan-Wigner mapping. It also includes the behaviour of the existing
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:31]: Is the output actually a `PauliSentence` object?

**Review by trbromley — COMMENTED** [review]
image
Why do we have 4 options and not 2?

File: pennylane/qchem/observable_hf.py [hunk 2]
```diff
@@ -140,85 +141,3 @@ def qubit_observable(o_ferm, cutoff=1.0e-12):
...
-    Returns:
-        tuple(list[complex], list[Operation]): list of coefficients and qubit operators
-
-    **Example**
-
-    >>> f  = [0, 0]
-    >>> q = jordan_wigner(f)
-    >>> q # corresponds to :math:`\frac{1}{2}(I_0 - Z_0)`
-    ([(0.5+0j), (-0.5+0j)], [Identity(wires=[0]), PauliZ(wires=[0])])
-    """
```
Comment (trbromley) [inline pennylane/qchem/observable_hf.py:166]: You could keep all of this (for now) and just have it return the top-level `jordan_wigner`

**Comment by lillian542** [comment]
> Why do we have 4 options and not 2?

One is the top-level function, one is FermiWord, one is FermiSentence, one is the old function

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (lillian542) [reply doc/releases/changelog-dev.md:31]: Ah, good point. Either PauliSentence or Operator, default is Operator.

**Comment by soranjh** [comment]
> Possible Drawbacks:
> The documentation for the old version is less accessible; the docstring for the new jordan_wigner doesn't mention it or explain its input format.

**Comment by soranjh** [comment]
> we could temporarily make the docstring for the existing jordan_wigner function longer and have it describe all 3 possible scenarios. 

The current docstrings are fine. The old `jordan_wigner` is just used internally in qchem.

File: pennylane/fermi/conversion.py [hunk 3]
```diff
@@ -126,3 +130,86 @@ def _(fermi_operator: FermiSentence, ps=False, wire_map=None):
         return qubit_operator.map_wires(wire_map)
 
     return qubit_operator
+
+
+@jordan_wigner.register
+def _jordan_wigner_legacy(op: list, notation="physicist"):  # pylint:disable=too-many-branches
```
Comment (soranjh) [inline pennylane/fermi/conversion.py:136]: Is anything in this function changed, except its name?

File: pennylane/fermi/conversion.py [hunk 4]
```diff
@@ -126,3 +130,86 @@ def _(fermi_operator: FermiSentence, ps=False, wire_map=None):
...
+            for j in k[::-1][:-1]:
+                del o[j]
+                c[k[0]] = c[k[0]] + c[j]
+                del c[j]
+
+    # Pauli gates objects pregenerated for speed
+    pauli_map = _get_pauli_map(np.max(op))
+    for i, term in enumerate(o):
+        if len(term) == 0:
+            o[i] = qml.Identity(0)
```
Comment (soranjh) [inline pennylane/fermi/conversion.py:210]: Any idea why this is not covered? The tests for the `qchem.jordan_wigner` are all in place?

**Review by soranjh — COMMENTED** [review]
Thanks @lillian542, looks good. Any idea why codecov complains about #L210?

**Comment by trbromley** [comment]
> > Why do we have 4 options and not 2?
> 
> One is the top-level function, one is FermiWord, one is FermiSentence, one is the old function

Thanks! But why do we have two different signatures for the first 3 options? Also, why do we need `ps`?

**Comment by lillian542** [comment]
> Thanks! But why do we have two different signatures for the first 3 options? Also, why do we need ps?

I'll update them to all take `fermionic_op` as the name of the first argument, so it's more similar. The overall function has to take `fermionic_op, **kwargs` because there are different kwargs for the new implementation and the old. Once we deprecate and remove the old version, the 3 remaining can all have the same signature.

We need the `ps` kwarg for if we want to convert to a `PauliSentence` instead of `Operator`. I think @soranjh or @Jaybsoni can say more specifically when we would want to use this functionality.

File: pennylane/fermi/conversion.py (hunk 3 above)
Comment (lillian542) [reply pennylane/fermi/conversion.py:136]: Name and location only, I copy-pasted from the `observable_hf` file to here and updated the name

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (lillian542) [inline doc/releases/changelog-dev.md:31]:
```suggestion
  `Operator` (or its equivalent `PauliSentence` based on an optional kwarg) using the Jordan-Wigner mapping. It also includes the behaviour of the existing
```

File: pennylane/fermi/conversion.py (hunk 4 above)
Comment (lillian542) [reply pennylane/fermi/conversion.py]: I just told codecov not to worry about it; doesn't seem worth it to figure out what is going on here if we are deprecating that version next release anyway.

**Comment by soranjh** [comment]
Also, why do we need ps?

Working with ps directly and not converting to operation makes some of the tapering functions more efficient.

File: pennylane/fermi/conversion.py (hunk 4 above)
Comment (soranjh) [reply pennylane/fermi/conversion.py]: Sounds good to me.

**Review by soranjh — APPROVED** [review]
Thanks @lillian542, performed some local tests, all seems good.

File: pennylane/fermi/conversion.py [hunk 5]
```diff
@@ -126,3 +130,90 @@ def _(fermi_operator: FermiSentence, ps=False, wire_map=None):
...
+
+    # Pauli gates objects pregenerated for speed
+    pauli_map = _get_pauli_map(np.max(op))
+    for i, term in enumerate(o):
+        if len(term) == 0:
+            # moved function from qchem.observable_hf without any changes to tests
+            # codecov complained this line is not covered
+            # function to be deprecated next release
+            # not going to write a test to cover this line
+            o[i] = qml.Identity(0)  # pragma: no-cover
```
Comment (soranjh) [inline pennylane/fermi/conversion.py:214]: Seems like a problem with codecov? 
WCS we can just remove this `if` condition, it seems that it is not needed probably due to some previous adjustments in qchem.

Comment (lillian542) [inline pennylane/fermi/conversion.py:214]:
```suggestion
            o[i] = qml.Identity(0)  # pragma: no cover
```

Comment (lillian542) [reply pennylane/fermi/conversion.py]: I just tagged it wrong, it should be `pragma: no cover` without the dash. Just updated, this time everything should pass 😅

File: pennylane/qchem/observable_hf.py (hunk 2 above)
Comment (lillian542) [reply pennylane/qchem/observable_hf.py]: done

File: pennylane/qchem/observable_hf.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def qubit_observable(o_ferm, cutoff=1.0e-12):
     return simplify(qml.Hamiltonian(coeffs, ops), cutoff=cutoff)
 
 
-def jordan_wigner(op, notation="physicist"):
+def jordan_wigner(op: list, notation="physicist"):  # pylint:disable=too-many-branches
```
Comment (Jaybsoni) [inline pennylane/qchem/observable_hf.py:145]: If the old jordan_wigner function was top level accessible, then I think it should be fine to just remove this completely! There may be some tests / parts of pennylane where we import it directly like `from pennylane.qchem import jordan_wigner`, but you can just find and replace all such instances with `qml.jordan_wigner`

Comment (lillian542) [reply pennylane/qchem/observable_hf.py:145]: It wasn't, it was only accessible as `qml.qchem.jordan_wigner`

**Comment by Jaybsoni** [comment]
> > image
> > Why do we have 4 options and not 2?
> 
> This issue can be fixed by single dispatching over a private function and wrapping it with the main function:
> 
> ```
> def jordan_wigner(fermi_operator, **kwargs):
>     """The main doc string"""
>     return _dispatched_jordan_wigner(fermi_operator, **kwargs)
> 
> @singledispatch
> def _dispatched_jordan_wigner(fermi_operator, *kwargs):
>     raise ValueError(f" ... ") 
> 
> @ _dispatched_jordan_wigner.register
> def _(fermi_operator: FermiWord, ps=False, wire_map=None):
>     ....
> ```

Spoke with @soranjh offline, we concluded that this can be left as is in this PR and then separately addressed in the documentation PR.

**Review by Jaybsoni — APPROVED** [review]

**Pull request merged**